<a href="https://colab.research.google.com/github/JALVARADORUIZ/Magister_BigData_UB/blob/main/M6_AI1_AlvaradoRuizJoseArmando.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

*PADAWAN*: 


##JOSE ARMANDO ALVARADO RUIZ

# **Tema 1. Árboles de decisiónArchivo**

**Entrega: RECAUDACIÓN DE PELÍCULAS (IMDB)**


# **Descripción de la tarea**

###Contexto

IMDb (acrónimo de Internet Movie Database) es una base de datos on-line de información relacionada con películas, programas de televisión, videos caseros, videojuegos y contenido de transmisión en línea - incluidos los actores, el equipo de producción y biografías personales, resúmenes de tramas, trivia, calificaciones y reseñas de fans y críticas. Originalmente, fue un sitio web operado por fanáticos, la base de datos ahora es propiedad y está operada por IMDb.com, Inc., una subsidiaria de Amazon.

A diciembre de 2020, IMDb tenía, aproximadamente, 7,5 millones de títulos (incluidos episodios) y 10,4 millones de personalidades en su base de datos, así como 83 millones de usuarios registrados. 
El dataset original es mucho mayor, pero ha sido reducido para eliminar procesos de limpieza de datos.

Descripción de las variables: 

* Imdb_title_id: id único de cada película. 
* Original_title: título original de la película. 
* Year: año. 
* Genre: género. 
* Duration: duración en minutos. 
* Country: país. 
* Director: director. 
* Actors: actor principal. 
* Avg_votes: media de los votos de los usuarios. 
* Avg_vote_weighted: media ponderada de la media de los votos con su número. 
*Votes: número de votos de los usuarios. 
*Budget: presupuesto destinado a su producción. 
*Worldwide_gross_income: ingresos generados en todo el mundo. 
*Reviews_from_users: número de opiniones de usuarios. 
*Reviews_from_critics: número de opiniones de críticos. 

Se pide: 

Aplicar el algoritmo de ÁRBOLES de DECISIÓN para la estimación de la recaudación de una película, dadas las variables: genre, duration, country, director, actors y Budget. 

Aplicar el algoritmo de ÁRBOLES de DECISIÓN para la estimación de la media ponderada de los votos en IMDB (avg_vote_weighted) atendiendo a las variables: genre, duration, country, director, actors, budget, worldwide_gross_income, reviews_from_users y reviews_from_critics.


### **Importación librerias**

In [84]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.tree import DecisionTreeClassifier
from sklearn.preprocessing import OrdinalEncoder

%matplotlib inline

# Carga de Datos y visualización general

In [85]:
file_name = '/content/act2_IMDb movies_ml.csv'
df = pd.read_csv(file_name, sep=';', decimal=',')

Visualización de los primeros 10 registros del dataset.

---




In [86]:
df.head(10)

imdb_title_id                       original_title  year  \
0     tt0010323         Das Cabinet des Dr. Caligari  1920   
1     tt0012190  The Four Horsemen of the Apocalypse  1921   
2     tt0014624    A Woman of Paris: A Drama of Fate  1923   
3     tt0015864                        The Gold Rush  1925   
4     tt0018455        Sunrise: A Song of Two Humans  1927   
5     tt0018578                                Wings  1927   
6     tt0021106                          Madam Satan  1930   
7     tt0021749                          City Lights  1931   
8     tt0021814                              Dracula  1931   
9     tt0022835              Dr. Jekyll and Mr. Hyde  1931   

                      genre  duration  country            director  \
0  Fantasy, Horror, Mystery        76  Germany        Robert Wiene   
1       Drama, Romance, War       150      USA          Rex Ingram   
2            Drama, Romance        82      USA     Charles Chaplin   
3  Adventure, Comedy, Drama        95      USA     Charles Chaplin   
4            Drama, Romance        94      USA         F.W. Murnau   
5       Drama, Romance, War       144      USA  William A. Wellman   
6    Comedy, Drama, Musical       116      USA    Cecil B. DeMille   
7    Comedy, Drama, Romance        87      USA     Charles Chaplin   
8    Drama, Fantasy, Horror        75      USA        Tod Browning   
9            Horror, Sci-Fi        98      USA    Rouben Mamoulian   

              actors avg_vote   votes avg_vote_weighted     budget  \
0      Werner Krauss      8.1   55601              7.99    $ 18000   
1     Pomeroy Cannon      7.2    3058              6.61   $ 800000   
2     Edna Purviance        7    4735              6.61   $ 351000   
3    Charles Chaplin      8.2   97480              8.13   $ 923000   
4     George O'Brien      8.1   46069              7.97   $ 200000   
5          Clara Bow      7.5   11139              7.18  $ 2000000   
6        Kay Johnson      6.5     838              6.11   $ 980000   
7  Virginia Cherrill      8.5  162668              8.45  $ 1500000   
8        Bela Lugosi      7.5   45054              7.41   $ 355000   
9      Fredric March      7.6   12225              7.28   $ 535000   

  worlwide_gross_income  reviews_from_users  reviews_from_critics  
0                $ 8811                 237                   160  
1             $ 9183673                  45                    16  
2               $ 11233                  37                    24  
3               $ 26916                 190                   125  
4              $ 121107                 250                   186  
5                 $ 746                  95                    92  
6                $ 1005                  43                    19  
7               $ 46008                 295                   122  
8               $ 85204                 556                   163  
9               $ 16615                 127                    62

In [87]:
# Estructura dataset
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8664 entries, 0 to 8663
Data columns (total 15 columns):
 #   Column                 Non-Null Count  Dtype 
---  ------                 --------------  ----- 
 0   imdb_title_id          8664 non-null   object
 1   original_title         8664 non-null   object
 2   year                   8664 non-null   int64 
 3   genre                  8664 non-null   object
 4   duration               8664 non-null   int64 
 5   country                8664 non-null   object
 6   director               8664 non-null   object
 7   actors                 8664 non-null   object
 8   avg_vote               8664 non-null   object
 9   votes                  8664 non-null   int64 
 10  avg_vote_weighted      8664 non-null   object
 11  budget                 8664 non-null   object
 12  worlwide_gross_income  8664 non-null   object
 13  reviews_from_users     8664 non-null   int64 
 14  reviews_from_critics   8664 non-null   int64 
dtypes: int64(5), object(1

In [88]:
# Analisis global de datos estadisticos de la muestra de variables cuantitativas
df.describe()

year     duration         votes  reviews_from_users  \
count  8664.000000  8664.000000  8.664000e+03         8664.000000   
mean   2002.270891   106.668629  7.464144e+04          251.428440   
std      13.389456    18.541422  1.502981e+05          454.674936   
min    1920.000000    75.000000  1.000000e+02            1.000000   
25%    1996.000000    94.000000  4.199250e+03           38.000000   
50%    2005.000000   103.000000  2.104800e+04          117.000000   
75%    2012.000000   116.000000  7.749050e+04          284.000000   
max    2020.000000   271.000000  2.278845e+06        10472.000000   

       reviews_from_critics  
count           8664.000000  
mean             114.249885  
std              121.558603  
min                1.000000  
25%               27.000000  
50%               73.000000  
75%              161.000000  
max              999.000000

El dataset consta de **8664 registros** con **15 variables**.

## **Consideraciones:**

Se deben realizar dos tipos de preprocesado: 

* Convertir todas las variables categóricas en numéricas. 
* Tanto “budget” como “worldwide_gross_income” están en formato string. Hay que eliminar el símbolo de dólar para convertir a entero. 

Con todos estos datos se pide lo siguiente: 

# Preprocesado de conversión de tipos de datos (comentado arriba).

In [89]:
# Creamos arreglos con valores unicos de las variables del tipo object:
list_actors = df.actors.unique()
list_directors = df.director.unique()
list_genres = df.genre.unique()
list_countries = df.country.unique()

In [90]:
# Creamos el codificador indicandole el orden de la variables. Se utiliza la funcion OrdinalEncoder de skit-learn
encoder_actors = OrdinalEncoder(categories=[list_actors])
encoder_directors = OrdinalEncoder(categories=[list_directors])
encoder_genres = OrdinalEncoder(categories=[list_genres])
encoder_countries = OrdinalEncoder(categories=[list_countries])

# Ajustamos el codificador con la variable education y la transformamos
encoder_actors.fit(df[["actors"]])
encoder_directors.fit(df[["director"]])
encoder_genres.fit(df[["genre"]])
encoder_countries.fit(df[["country"]])

# Genero nuevas variables con la asignación numerica de las variables categoricas:

df["actors-encoded"] = encoder_actors.transform(df[["actors"]])
df["director-encoded"] = encoder_directors.transform(df[["director"]])
df["genre-encoded"] = encoder_genres.transform(df[["genre"]])
df["country-encoded"] = encoder_countries.transform(df[["country"]])

In [91]:
df[["actors", "actors-encoded","director","director-encoded","genre","genre-encoded","country","country-encoded"]].head(10)

actors  actors-encoded            director  director-encoded  \
0      Werner Krauss             0.0        Robert Wiene               0.0   
1     Pomeroy Cannon             1.0          Rex Ingram               1.0   
2     Edna Purviance             2.0     Charles Chaplin               2.0   
3    Charles Chaplin             3.0     Charles Chaplin               2.0   
4     George O'Brien             4.0         F.W. Murnau               3.0   
5          Clara Bow             5.0  William A. Wellman               4.0   
6        Kay Johnson             6.0    Cecil B. DeMille               5.0   
7  Virginia Cherrill             7.0     Charles Chaplin               2.0   
8        Bela Lugosi             8.0        Tod Browning               6.0   
9      Fredric March             9.0    Rouben Mamoulian               7.0   

                      genre  genre-encoded  country  country-encoded  
0  Fantasy, Horror, Mystery            0.0  Germany              0.0  
1       Drama, Romance, War            1.0      USA              1.0  
2            Drama, Romance            2.0      USA              1.0  
3  Adventure, Comedy, Drama            3.0      USA              1.0  
4            Drama, Romance            2.0      USA              1.0  
5       Drama, Romance, War            1.0      USA              1.0  
6    Comedy, Drama, Musical            4.0      USA              1.0  
7    Comedy, Drama, Romance            5.0      USA              1.0  
8    Drama, Fantasy, Horror            6.0      USA              1.0  
9            Horror, Sci-Fi            7.0      USA              1.0

In [92]:
# Conversion de variables “budget” como “worldwide_gross_income”

# Elimino simbolo $ mediante expresion regular y replace. Se convierte columna en int:
df.budget = df.budget.str.replace(r"^[$\s]+", '',regex=True).astype(int)
df.worlwide_gross_income = df.worlwide_gross_income.str.replace(r"^[$\s]+", '',regex=True).astype(int)

In [93]:
# Convierto variables object a numeric (float)
df.avg_vote_weighted = df.avg_vote_weighted.astype(float)
df.avg_vote = df.avg_vote.astype(float)

In [94]:
# Estructura de dataframe luego de preprocesado.

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8664 entries, 0 to 8663
Data columns (total 19 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   imdb_title_id          8664 non-null   object 
 1   original_title         8664 non-null   object 
 2   year                   8664 non-null   int64  
 3   genre                  8664 non-null   object 
 4   duration               8664 non-null   int64  
 5   country                8664 non-null   object 
 6   director               8664 non-null   object 
 7   actors                 8664 non-null   object 
 8   avg_vote               8664 non-null   float64
 9   votes                  8664 non-null   int64  
 10  avg_vote_weighted      8664 non-null   float64
 11  budget                 8664 non-null   int64  
 12  worlwide_gross_income  8664 non-null   int64  
 13  reviews_from_users     8664 non-null   int64  
 14  reviews_from_critics   8664 non-null   int64  
 15  acto

In [95]:
df.head(10)

imdb_title_id                       original_title  year  \
0     tt0010323         Das Cabinet des Dr. Caligari  1920   
1     tt0012190  The Four Horsemen of the Apocalypse  1921   
2     tt0014624    A Woman of Paris: A Drama of Fate  1923   
3     tt0015864                        The Gold Rush  1925   
4     tt0018455        Sunrise: A Song of Two Humans  1927   
5     tt0018578                                Wings  1927   
6     tt0021106                          Madam Satan  1930   
7     tt0021749                          City Lights  1931   
8     tt0021814                              Dracula  1931   
9     tt0022835              Dr. Jekyll and Mr. Hyde  1931   

                      genre  duration  country            director  \
0  Fantasy, Horror, Mystery        76  Germany        Robert Wiene   
1       Drama, Romance, War       150      USA          Rex Ingram   
2            Drama, Romance        82      USA     Charles Chaplin   
3  Adventure, Comedy, Drama        95      USA     Charles Chaplin   
4            Drama, Romance        94      USA         F.W. Murnau   
5       Drama, Romance, War       144      USA  William A. Wellman   
6    Comedy, Drama, Musical       116      USA    Cecil B. DeMille   
7    Comedy, Drama, Romance        87      USA     Charles Chaplin   
8    Drama, Fantasy, Horror        75      USA        Tod Browning   
9            Horror, Sci-Fi        98      USA    Rouben Mamoulian   

              actors  avg_vote   votes  avg_vote_weighted   budget  \
0      Werner Krauss       8.1   55601               7.99    18000   
1     Pomeroy Cannon       7.2    3058               6.61   800000   
2     Edna Purviance       7.0    4735               6.61   351000   
3    Charles Chaplin       8.2   97480               8.13   923000   
4     George O'Brien       8.1   46069               7.97   200000   
5          Clara Bow       7.5   11139               7.18  2000000   
6        Kay Johnson       6.5     838               6.11   980000   
7  Virginia Cherrill       8.5  162668               8.45  1500000   
8        Bela Lugosi       7.5   45054               7.41   355000   
9      Fredric March       7.6   12225               7.28   535000   

   worlwide_gross_income  reviews_from_users  reviews_from_critics  \
0                   8811                 237                   160   
1                9183673                  45                    16   
2                  11233                  37                    24   
3                  26916                 190                   125   
4                 121107                 250                   186   
5                    746                  95                    92   
6                   1005                  43                    19   
7                  46008                 295                   122   
8                  85204                 556                   163   
9                  16615                 127                    62   

   actors-encoded  director-encoded  genre-encoded  country-encoded  
0             0.0               0.0            0.0              0.0  
1             1.0               1.0            1.0              1.0  
2             2.0               2.0            2.0              1.0  
3             3.0               2.0            3.0              1.0  
4             4.0               3.0            2.0              1.0  
5             5.0               4.0            1.0              1.0  
6             6.0               5.0            4.0              1.0  
7             7.0               2.0            5.0              1.0  
8             8.0               6.0            6.0              1.0  
9             9.0               7.0            7.0              1.0

# Generación de los dos modelos que se piden. 


# Comentarios sobre la elección de los parámetros de los modelos finales. 


* Comentarios sobre los resultados. 


# Estimación del presupuesto de dos películas con los valores que queráis; eligiendo país, género, director, etc., a vuestro gusto. 


# Estimación del rating de dos películas con los valores que queráis; eligiendo país, género, director, etc., a vuestro gusto. 